File: relayer/relayer.go
package relayer

import (
	"context"
	"path/filepath"
	"strings"
	"time"

	rpchttp "github.com/cometbft/cometbft/rpc/client/http"

	"github.com/cosmos/cosmos-sdk/client"
	"github.com/cosmos/cosmos-sdk/server"
	sdk "github.com/cosmos/cosmos-sdk/types"
	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"

	"steemvm/x/steembridge/types"
)

// Start runs the in-process Steem relayer until ctx is canceled. It is
// spawned as a goroutine by the start command and MUST never crash the node:
// every failure path logs and retries. It returns immediately when the
// [steem-relayer] app.toml section is not configured.
func Start(ctx context.Context, svrCtx *server.Context, clientCtx client.Context) {
	logger := svrCtx.Logger.With("module", "steem-relayer")

	cfg := ReadFromAppOpts(svrCtx.Viper)
	if !cfg.Enabled() {
		logger.Info("steem relayer disabled (set [steem-relayer] steem-rpc-url and key-name in app.toml to enable)")
		return
	}

	// Resolve the signing key from the node's keyring.
	if clientCtx.Keyring == nil {
		logger.Error("steem relayer: no keyring available on client context")
		return
	}
	record, err := clientCtx.Keyring.Key(cfg.KeyName)
	if err != nil {
		logger.Error("steem relayer: key not found in keyring", "key", cfg.KeyName, "err", err)
		return
	}
	signerAddr, err := record.GetAddress()
	if err != nil {
		logger.Error("steem relayer: cannot read key address", "err", err)
		return
	}
	valoperAddr := sdk.ValAddress(signerAddr).String()

	// Connect to the local node RPC.
	rpcClient, err := rpchttp.New(svrCtx.Config.RPC.ListenAddress, "/websocket")
	if err != nil {
		logger.Error("steem relayer: cannot create local RPC client", "err", err)
		return
	}
	clientCtx = clientCtx.
		WithClient(rpcClient).
		WithFromName(cfg.KeyName).
		WithFromAddress(signerAddr).
		WithBroadcastMode("sync")

	// Wait until the local node is up and synced, then learn the chain-id.
	chainID := waitForLocalNode(ctx, logger, rpcClient)
	if chainID == "" {
		return // ctx canceled
	}
	clientCtx = clientCtx.WithChainID(chainID)

	steem := NewSteemClient(cfg.SteemRPCURL)
	bridgeQuery := types.NewQueryClient(clientCtx)
	stakingQuery := stakingtypes.NewQueryClient(clientCtx)
	stateDir := filepath.Join(svrCtx.Config.RootDir, "data")

	state, err := LoadState(stateDir)
	if err != nil {
		logger.Error("steem relayer: cannot load state", "err", err)
		return
	}

	logger.Info("steem relayer started",
		"steem_rpc", cfg.SteemRPCURL, "key", cfg.KeyName, "signer", signerAddr.String(),
		"poll_interval", cfg.PollInterval.String(), "last_scanned_block", state.LastScannedBlock)

	ticker := time.NewTicker(cfg.PollInterval)
	defer ticker.Stop()
	notBondedLogged := false

	for {
		select {
		case <-ctx.Done():
			logger.Info("steem relayer stopped")
			return
		case <-ticker.C:
		}

		if err := runCycle(ctx, logger, cfg, clientCtx, steem, bridgeQuery, stakingQuery, signerAddr, valoperAddr, stateDir, &state, &notBondedLogged); err != nil {
			logger.Error("steem relayer cycle failed", "err", err)
		}
	}
}

type cycleLogger interface {
	Info(msg string, keyVals ...any)
	Error(msg string, keyVals ...any)
	Debug(msg string, keyVals ...any)
}

// runCycle performs one poll: check feature flags and bonded status, scan
// new irreversible Steem blocks, attest gateway transfers, advance cursor.
func runCycle(
	ctx context.Context,
	logger cycleLogger,
	cfg Config,
	clientCtx client.Context,
	steem *SteemClient,
	bridgeQuery types.QueryClient,
	stakingQuery stakingtypes.QueryClient,
	signerAddr sdk.AccAddress,
	valoperAddr string,
	stateDir string,
	state *State,
	notBondedLogged *bool,
) error {
	params, err := bridgeQuery.Params(ctx, &types.QueryParamsRequest{})
	if err != nil {
		return err
	}
	gateway := params.Params.GatewayAccount
	if gateway == "" || (!params.Params.BridgeEnabled && !params.Params.NameServiceEnabled) {
		logger.Debug("steem relayer idle: bridge and name service disabled or no gateway configured")
		return nil
	}

	// Only bonded validators' attestations count (and only theirs are
	// fee-exempt) — idle quietly otherwise.
	valResp, err := stakingQuery.Validator(ctx, &stakingtypes.QueryValidatorRequest{ValidatorAddr: valoperAddr})
	if err != nil || !valResp.Validator.IsBonded() {
		if !*notBondedLogged {
			logger.Info("steem relayer idle: key is not a bonded validator", "valoper", valoperAddr)
			*notBondedLogged = true
		}
		return nil
	}
	*notBondedLogged = false

	lib, err := steem.LastIrreversibleBlock()
	if err != nil {
		return err
	}

	// First run: establish the cursor. Precedence: the node's local
	// start-block override, then the chain-wide relayer_start_block param
	// (the shared anchor recorded at launch so every validator scans the
	// same range regardless of when it first starts), then the current LIB.
	if state.LastScannedBlock == 0 {
		state.LastScannedBlock = initialCursor(cfg.StartBlock, params.Params.RelayerStartBlock, lib)
		logger.Info("steem relayer cursor initialized",
			"cursor", state.LastScannedBlock,
			"source", cursorSource(cfg.StartBlock, params.Params.RelayerStartBlock))
		return SaveState(stateDir, *state)
	}

	from := state.LastScannedBlock + 1
	to := min(lib, state.LastScannedBlock+cfg.MaxBlocksPerPoll)
	if from > to {
		return nil
	}

	blocks, err := steem.FetchBlocks(from, to)
	if err != nil {
		return err
	}

	// Accumulate attestations across blocks, keeping block boundaries so the
	// cursor only advances past fully-included blocks. One tx per cycle.
	var msgs []sdk.Msg
	lastFullBlock := state.LastScannedBlock
	for _, nb := range blocks {
		transfers := ExtractGatewayTransfers(nb.Num, nb.Block, gateway)
		blockMsgs := make([]sdk.Msg, 0, len(transfers))
		for _, transfer := range transfers {
			intent := RouteMemo(transfer.Memo)
			if intent == IntentRegister && !params.Params.NameServiceEnabled {
				continue
			}
			if intent == IntentDeposit && !params.Params.BridgeEnabled {
				continue
			}
			attested, err := alreadyAttested(ctx, bridgeQuery, transfer, intent, valoperAddr)
			if err != nil {
				return err
			}
			if attested {
				continue
			}
			blockMsgs = append(blockMsgs, BuildMsg(transfer, intent, signerAddr, gateway))
		}
		if len(msgs)+len(blockMsgs) > MaxMsgsPerTx {
			break // stop at a block boundary; the rest is picked up next cycle
		}
		msgs = append(msgs, blockMsgs...)
		lastFullBlock = nb.Num
	}

	if len(msgs) > 0 {
		txHash, err := BroadcastAttestations(ctx, clientCtx, cfg.KeyName, msgs)
		if err != nil {
			return err // cursor untouched; the same blocks are rescanned next cycle
		}
		logger.Info("steem relayer attested transfers",
			"count", len(msgs), "blocks", lastFullBlock-state.LastScannedBlock, "tx", txHash)
	}

	if lastFullBlock != state.LastScannedBlock {
		state.LastScannedBlock = lastFullBlock
		return SaveState(stateDir, *state)
	}
	return nil
}

// initialCursor picks the first-run scan cursor (the last "already scanned"
// block, i.e. scanning begins at cursor+1): a local app.toml start-block
// override wins, then the chain's relayer_start_block param, then Steem's
// current last irreversible block.
func initialCursor(localStart, chainStart, lib uint64) uint64 {
	switch {
	case localStart > 0:
		return localStart - 1
	case chainStart > 0:
		return chainStart - 1
	default:
		return lib
	}
}

func cursorSource(localStart, chainStart uint64) string {
	switch {
	case localStart > 0:
		return "app.toml start-block override"
	case chainStart > 0:
		return "on-chain relayer_start_block param"
	default:
		return "current steem last irreversible block"
	}
}

// alreadyAttested reports whether this validator has already confirmed the
// transfer's (txid, opIndex) key, or the record is past the point where new
// attestations are accepted. Keeps rescans (lost state file, retried cycles)
// idempotent — a duplicate message would disqualify the whole zero-fee tx.
func alreadyAttested(ctx context.Context, q types.QueryClient, t Transfer, intent Intent, valoperAddr string) (bool, error) {
	switch intent {
	case IntentRegister:
		resp, err := q.NameRegistrationByTxid(ctx, &types.QueryNameRegistrationByTxidRequest{Txid: t.Txid, OpIndex: t.OpIndex})
		if err != nil {
			return false, ignoreNotFound(err)
		}
		if resp.Registration.Status != types.NameRegistrationStatus_NAME_REGISTRATION_STATUS_PENDING {
			return true, nil
		}
		return hasConfirmation(resp.Registration.ValidatorConfirmations, valoperAddr), nil
	default:
		resp, err := q.DepositByTxid(ctx, &types.QueryDepositByTxidRequest{Txid: t.Txid, OpIndex: t.OpIndex})
		if err != nil {
			return false, ignoreNotFound(err)
		}
		if resp.Deposit.Status != types.DepositStatus_DEPOSIT_STATUS_PENDING {
			return true, nil
		}
		return hasConfirmation(resp.Deposit.ValidatorConfirmations, valoperAddr), nil
	}
}

func hasConfirmation(confirmations []*types.Confirmation, valoperAddr string) bool {
	for _, c := range confirmations {
		if c.ValidatorAddress == valoperAddr {
			return true
		}
	}
	return false
}

// ignoreNotFound maps "key not found" query errors to nil (meaning: not yet
// attested); anything else is a real error that aborts the cycle. ABCI query
// errors arrive as flattened strings, so this matches on the message.
func ignoreNotFound(err error) error {
	if err == nil {
		return nil
	}
	if strings.Contains(err.Error(), "not found") || strings.Contains(err.Error(), "NotFound") {
		return nil
	}
	return err
}

// waitForLocalNode blocks until the local node RPC responds and is done
// catching up, returning the chain-id (empty when ctx is canceled).
func waitForLocalNode(ctx context.Context, logger cycleLogger, rpcClient *rpchttp.HTTP) string {
	for {
		select {
		case <-ctx.Done():
			return ""
		case <-time.After(2 * time.Second):
		}
		status, err := rpcClient.Status(ctx)
		if err != nil {
			continue // node still booting
		}
		if status.SyncInfo.CatchingUp {
			logger.Debug("steem relayer waiting: node is catching up")
			continue
		}
		return status.NodeInfo.Network
	}
}

File: oracle/js/src/relayer.ts
// relayer.ts — the main scan/attest loop, mirroring oracle/go/relayer/relayer.go
// field-for-field. Uses the REST gRPC-gateway (:1317) for every chain query
// (params, staking validator lookup, dedup pre-checks) — see
// oracle/PROTOCOL.md §4's recommendation for Python/JS over raw CometBFT
// RPC/gRPC.

import { toBech32, fromBech32 } from "@cosmjs/encoding";
import type { Registry } from "@cosmjs/proto-signing";

import type { Config } from "./config";
import type { EthSecp256k1DirectSigner } from "./signer";
import { SteemClient, extractGatewayTransfers, extractGatewayPayouts } from "./steemClient";
import { routeMemo, buildMsg, Intent, GATEWAY_ACCOUNT } from "./router";
import { broadcastAttestations, broadcastPriceFeedMsgs, type EncodeObject } from "./broadcast";
import { loadState, saveState, loadFeederState, saveFeederState } from "./state";
import { Feeder, getAggregateVoteHash, type PriceSource } from "./priceFeeder";
import { TYPE_URL_MSG_ATTEST_WITHDRAWAL_PAYOUT } from "./broadcast";

export interface CycleLogger {
  info(msg: string, meta?: Record<string, unknown>): void;
  error(msg: string, meta?: Record<string, unknown>): void;
  debug(msg: string, meta?: Record<string, unknown>): void;
}

const VALOPER_HRP = "steemvaloper";

/** Derives the bech32 valoper address from an account bech32 address — same
 * 20 bytes, different HRP (see oracle/PROTOCOL.md §1). Only used for
 * `x/staking` bonded-status lookups; every duty message field uses the
 * account address. */
function toValoperAddress(accountAddress: string): string {
  const { data } = fromBech32(accountAddress);
  return toBech32(VALOPER_HRP, data);
}

async function getJson<T>(restUrl: string, path: string): Promise<T> {
  const resp = await fetch(`${trimSlash(restUrl)}${path}`);
  if (resp.status === 404) {
    const err = new Error(`not found: ${path}`);
    (err as any).notFound = true;
    throw err;
  }
  if (!resp.ok) {
    throw new Error(`GET ${path}: HTTP ${resp.status}: ${await resp.text().catch(() => "")}`);
  }
  return resp.json() as Promise<T>;
}

function trimSlash(url: string): string {
  return url.endsWith("/") ? url.slice(0, -1) : url;
}

interface BridgeParams {
  gateway_account: string;
  bridge_enabled: boolean;
  name_service_enabled: boolean;
  bridge_out_enabled: boolean;
  relayer_start_block: string;
}

async function queryBridgeParams(restUrl: string): Promise<BridgeParams> {
  const resp = await getJson<{ params: BridgeParams }>(restUrl, "/steemvm/steembridge/v1/params");
  return resp.params;
}

async function queryValidatorBonded(restUrl: string, valoperAddr: string): Promise<boolean> {
  try {
    const resp = await getJson<{ validator: { status: string } }>(
      restUrl,
      `/cosmos/staking/v1beta1/validators/${valoperAddr}`,
    );
    return resp.validator.status === "BOND_STATUS_BONDED";
  } catch (err) {
    if ((err as any).notFound) return false;
    throw err;
  }
}

async function depositAlreadyAttested(
  restUrl: string,
  txid: string,
  opIndex: number,
  valoperAddr: string,
): Promise<boolean> {
  try {
    const resp = await getJson<{
      deposit: { status: string; validator_confirmations: { validator_address: string }[] };
    }>(restUrl, `/steemvm/steembridge/v1/deposit_by_txid/${encodeURIComponent(txid)}/${opIndex}`);
    if (resp.deposit.status !== "DEPOSIT_STATUS_PENDING") return true;
    return resp.deposit.validator_confirmations.some((c) => c.validator_address === valoperAddr);
  } catch (err) {
    if ((err as any).notFound) return false;
    throw err;
  }
}

async function nameRegistrationAlreadyAttested(
  restUrl: string,
  txid: string,
  opIndex: number,
  valoperAddr: string,
): Promise<boolean> {
  try {
    const resp = await getJson<{
      registration: { status: string; validator_confirmations: { validator_address: string }[] };
    }>(restUrl, `/steemvm/steembridge/v1/name_registration_by_txid/${encodeURIComponent(txid)}/${opIndex}`);
    if (resp.registration.status !== "NAME_REGISTRATION_STATUS_PENDING") return true;
    return resp.registration.validator_confirmations.some((c) => c.validator_address === valoperAddr);
  } catch (err) {
    if ((err as any).notFound) return false;
    throw err;
  }
}

async function queryOracleDataParams(
  restUrl: string,
): Promise<{ vote_period: string; whitelist: string[] }> {
  const resp = await getJson<{ params: { vote_period: string; whitelist: string[] } }>(
    restUrl,
    "/steemvm/oracle/data/v1/params",
  );
  return resp.params;
}

async function queryLatestHeight(restUrl: string): Promise<number> {
  const resp = await getJson<{ block: { header: { height: string } } }>(
    restUrl,
    "/cosmos/base/tendermint/v1beta1/blocks/latest",
  );
  return Number(resp.block.header.height);
}

/** Waits until the node's REST API is reachable and (best-effort) not
 * catching up, returning the chain-id read from node_info. */
async function waitForLocalNode(logger: CycleLogger, restUrl: string): Promise<string> {
  for (;;) {
    try {
      const [nodeInfo, syncing] = await Promise.all([
        getJson<{ default_node_info: { network: string } }>(restUrl, "/cosmos/base/tendermint/v1beta1/node_info"),
        getJson<{ syncing: boolean }>(restUrl, "/cosmos/base/tendermint/v1beta1/syncing").catch(() => ({
          syncing: false,
        })),
      ]);
      if (syncing.syncing) {
        logger.debug("steem relayer waiting: node is catching up");
        await sleep(2_000);
        continue;
      }
      return nodeInfo.default_node_info.network;
    } catch {
      await sleep(2_000);
    }
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/** initialCursor picks the first-run scan cursor (the last "already
 * scanned" block, i.e. scanning begins at cursor+1): a local start-block
 * override wins, then the chain's relayer_start_block param, then Steem's
 * current last irreversible block. */
function initialCursor(localStart: number, localStartIsLatest: boolean, lib: number, chainStart: number): number {
  if (localStartIsLatest) return lib;
  if (localStart > 0) return localStart - 1;
  if (chainStart > 0) return chainStart - 1;
  return lib;
}

/**
 * Drives the whole oracle: bridge attestation scanning + (optionally) the
 * price feeder, on cfg.pollIntervalMs, until abortSignal fires. Mirrors
 * oracle/go/relayer/relayer.go's Run.
 */
export async function run(opts: {
  logger: CycleLogger;
  cfg: Config;
  signer: EthSecp256k1DirectSigner;
  registry: Registry;
  stateDir: string;
  priceSource: PriceSource | undefined;
  abortSignal: AbortSignal;
}): Promise<void> {
  const { logger, cfg, signer, registry, stateDir, priceSource, abortSignal } = opts;

  const valoperAddr = toValoperAddress(signer.address);
  const steem = new SteemClient(cfg.steemRpcUrl);

  const chainId = await waitForLocalNode(logger, cfg.nodeRestUrl);
  if (abortSignal.aborted) return;

  let state = await loadState(stateDir);
  const feeder = new Feeder(signer.address, priceSource);
  let lastHandledPeriod = 0;

  logger.info("steem oracle started", {
    steem_rpc: cfg.steemRpcUrl,
    node_rest: cfg.nodeRestUrl,
    chain_id: chainId,
    signer: signer.address,
    valoper: valoperAddr,
    poll_interval_ms: cfg.pollIntervalMs,
    last_scanned_block: state.last_scanned_block,
    price_feeder_enabled: priceSource !== undefined,
  });

  let notBondedLogged = false;
  // Heartbeat: the "idle"/"waiting" logs below are debug-level (filtered out
  // by default), so a quiet cycle -- no new transfers to attest, which is
  // most cycles most of the time -- produces zero output at all without
  // this. Mirrors oracle/go/relayer/relayer.go's heartbeat.
  const HEARTBEAT_EVERY = 10;
  let tickCount = 0;

  while (!abortSignal.aborted) {
    await sleepOrAbort(cfg.pollIntervalMs, abortSignal);
    if (abortSignal.aborted) break;

    try {
      const result = await runCycle(logger, cfg, chainId, signer, registry, steem, stateDir, state, valoperAddr, notBondedLogged);
      state = result.state;
      notBondedLogged = result.notBondedLogged;
    } catch (err) {
      logger.error("steem oracle cycle failed", { err: String(err) });
    }

    if (priceSource) {
      try {
        lastHandledPeriod = await runPriceFeederCycle(
          logger,
          cfg,
          chainId,
          signer,
          registry,
          feeder,
          stateDir,
          lastHandledPeriod,
        );
      } catch (err) {
        logger.error("price feeder cycle failed", { err: String(err) });
      }
    }

    tickCount++;
    if (tickCount % HEARTBEAT_EVERY === 0) {
      logger.info("steem oracle heartbeat", { last_scanned_block: state.last_scanned_block });
    }
  }

  logger.info("steem oracle stopped");
}

async function sleepOrAbort(ms: number, signal: AbortSignal): Promise<void> {
  if (signal.aborted) return;
  await new Promise<void>((resolve) => {
    const t = setTimeout(resolve, ms);
    signal.addEventListener("abort", () => {
      clearTimeout(t);
      resolve();
    }, { once: true });
  });
}

interface RelayerCycleState {
  last_scanned_block: number;
}

async function runCycle(
  logger: CycleLogger,
  cfg: Config,
  chainId: string,
  signer: EthSecp256k1DirectSigner,
  registry: Registry,
  steem: SteemClient,
  stateDir: string,
  state: RelayerCycleState,
  valoperAddr: string,
  notBondedLogged: boolean,
): Promise<{ state: RelayerCycleState; notBondedLogged: boolean }> {
  const params = await queryBridgeParams(cfg.nodeRestUrl);
  // GatewayAccount is a hardcoded chain constant, never read from params —
  // see router.GATEWAY_ACCOUNT's doc comment.
  const gateway = GATEWAY_ACCOUNT;
  if (!params.bridge_enabled && !params.name_service_enabled) {
    logger.debug("steem relayer idle: bridge and name service disabled");
    return { state, notBondedLogged };
  }

  const bonded = await queryValidatorBonded(cfg.nodeRestUrl, valoperAddr);
  if (!bonded) {
    if (!notBondedLogged) {
      logger.info("steem relayer idle: key is not a bonded validator", { valoper: valoperAddr });
    }
    return { state, notBondedLogged: true };
  }
  notBondedLogged = false;

  const lib = await steem.lastIrreversibleBlock();

  if (state.last_scanned_block === 0) {
    const cursor = initialCursor(cfg.startBlock, cfg.startBlockIsLatest, lib, Number(params.relayer_start_block));
    const next = { last_scanned_block: cursor };
    logger.info("steem relayer cursor initialized", { cursor });
    await saveState(stateDir, next);
    return { state: next, notBondedLogged };
  }

  const from = state.last_scanned_block + 1;
  const to = Math.min(lib, state.last_scanned_block + cfg.maxBlocksPerPoll);
  if (from > to) {
    return { state, notBondedLogged };
  }

  const blocks = await steem.fetchBlocks(from, to);

  const msgs: EncodeObject[] = [];
  let lastFullBlock = state.last_scanned_block;

  for (const nb of blocks) {
    const transfers = extractGatewayTransfers(nb.num, nb.block, gateway, cfg.steemSymbol, cfg.sbdSymbol);
    const blockMsgs: EncodeObject[] = [];

    for (const transfer of transfers) {
      const intent = routeMemo(transfer.memo);
      if (intent === Intent.Register && !params.name_service_enabled) continue;
      if (intent === Intent.Deposit && !params.bridge_enabled) continue;

      // Only attest transfers a claimable-destination memo parser would
      // accept, mirroring the chain's own DeriveDestination check: a bare
      // steem1.../0x address optionally prefixed svm-deposit/svm-register.
      if (!hasParseableDestination(transfer.memo)) continue;

      const attested =
        intent === Intent.Register
          ? await nameRegistrationAlreadyAttested(cfg.nodeRestUrl, transfer.txid, transfer.opIndex, valoperAddr)
          : await depositAlreadyAttested(cfg.nodeRestUrl, transfer.txid, transfer.opIndex, valoperAddr);
      if (attested) continue;

      blockMsgs.push(buildMsg(transfer, intent, signer.address, gateway));
      logger.info("attesting transfer", {
        intent: intent === Intent.Register ? "name-registration" : "deposit",
        txid: transfer.txid,
        opIndex: transfer.opIndex,
        from: transfer.from,
        amountMillisteem: transfer.amountMillisteem.toString(),
        memo: transfer.memo,
      });
    }

    if (params.bridge_out_enabled) {
      for (const payout of extractGatewayPayouts(nb.num, nb.block, gateway)) {
        blockMsgs.push({
          typeUrl: TYPE_URL_MSG_ATTEST_WITHDRAWAL_PAYOUT,
          value: {
            validator: signer.address,
            withdrawalId: payout.withdrawalId.toString(),
            steemTxid: payout.txid,
            opIndex: payout.opIndex,
            steemBlock: payout.steemBlock,
            steemTimestamp: payout.steemTimestamp,
          },
        });
        logger.info("attesting withdrawal payout", {
          withdrawalId: payout.withdrawalId.toString(),
          steemTxid: payout.txid,
          opIndex: payout.opIndex,
        });
      }
    }

    if (msgs.length + blockMsgs.length > 50) {
      break; // stop at a block boundary; the rest is picked up next cycle
    }
    msgs.push(...blockMsgs);
    lastFullBlock = nb.num;
  }

  if (msgs.length > 0) {
    const result = await broadcastAttestations({ restUrl: cfg.nodeRestUrl, chainId, signer, registry, msgs });
    if (result.code !== 0) {
      throw new Error(`attestation tx rejected (code ${result.code}): ${result.rawLog}`);
    }
    logger.info("steem relayer attested transfers", {
      count: msgs.length,
      blocks: lastFullBlock - state.last_scanned_block,
      tx: result.txHash,
    });
  }

  if (lastFullBlock !== state.last_scanned_block) {
    const next = { last_scanned_block: lastFullBlock };
    await saveState(stateDir, next);
    return { state: next, notBondedLogged };
  }
  return { state, notBondedLogged };
}

const HEX_ADDRESS_RE = /^0x[0-9a-fA-F]{40}$/;

// hasParseableDestination mirrors x/oracle/bridge/types/memo.go's
// DeriveDestination exactly enough to decide whether a transfer is worth
// attesting at all (the Go relayer applies the identical gate in
// relayer.go's runCycle) — full bech32 checksum validation via
// @cosmjs/encoding's fromBech32 (not just a shape regex), same
// single-leading-prefix stripping rule, same 0x-EVM-address fallback. The
// chain re-derives the destination itself at resolution time regardless —
// this is purely a "don't waste a broadcast on an obviously unparseable
// memo" pre-filter, never the source of truth.
function hasParseableDestination(memo: string): boolean {
  let trimmed = memo.trim();

  for (const prefix of ["svm-deposit", "svm-register"]) {
    if (trimmed.startsWith(prefix)) {
      const rest = trimmed.slice(prefix.length);
      if (rest === "" || rest[0] === " " || rest[0] === "\t") {
        trimmed = rest.trim();
      }
      break;
    }
  }

  try {
    const { prefix } = fromBech32(trimmed);
    if (prefix === ACCOUNT_HRP_FOR_MEMO) return true;
  } catch {
    // not a valid bech32 string at all — fall through to the hex check
  }

  return HEX_ADDRESS_RE.test(trimmed);
}

// Local constant to avoid importing signer.ts's ACCOUNT_HRP just for this
// one check (relayer.ts already imports enough from signer.ts's siblings);
// kept in sync with app/config.go's account bech32 prefix "steem".
const ACCOUNT_HRP_FOR_MEMO = "steem";

async function runPriceFeederCycle(
  logger: CycleLogger,
  cfg: Config,
  chainId: string,
  signer: EthSecp256k1DirectSigner,
  registry: Registry,
  feeder: Feeder,
  stateDir: string,
  lastHandledPeriod: number,
): Promise<number> {
  const params = await queryOracleDataParams(cfg.nodeRestUrl);
  const votePeriod = Number(params.vote_period);
  if (votePeriod === 0) {
    throw new Error("oracledata vote period is zero");
  }

  const height = await queryLatestHeight(cfg.nodeRestUrl);
  const period = Math.floor(height / votePeriod);
  if (period === lastHandledPeriod) {
    return lastHandledPeriod;
  }

  const prevState = await loadFeederState(stateDir);
  const { msgs, state: newState } = await feeder.step(period, params.whitelist, prevState);
  if (msgs.length === 0) {
    return period;
  }

  if (prevState.exchange_rates !== "" && prevState.prevote_period + 1 === period) {
    logger.info("submitting price vote", { period, exchangeRates: prevState.exchange_rates });
  }
  if (newState.exchange_rates !== "") {
    // Logs the hash, not the rates themselves -- mirrors the commit-reveal
    // scheme's own hide-until-reveal semantics (oracle/go/relayer does the
    // same), even though this is a local-only log with no other observer.
    logger.info("submitting price prevote", {
      period,
      hash: getAggregateVoteHash(newState.salt, newState.exchange_rates, signer.address),
    });
  }

  const result = await broadcastPriceFeedMsgs({
    restUrl: cfg.nodeRestUrl,
    chainId,
    signer,
    registry,
    msgs,
    gasPrices: cfg.gasPrices,
  });
  if (result.code !== 0) {
    throw new Error(`price feed tx rejected (code ${result.code}): ${result.rawLog}`);
  }
  logger.info("price feeder broadcast", { period, count: msgs.length, tx: result.txHash });

  await saveFeederState(stateDir, newState);
  return period;
}
